src/otaclient/ota_core.py

```diff
@@ -427,6 +408,16 @@ def _execute_update(self):
         """Implementation of OTA updating."""
         logger.info(f"execute local update({ecu_info.ecu_id=}): {self.update_version=}")
 
+        if _upper_proxy := self._upper_proxy:
+            logger.info(
+                f"use {_upper_proxy} for local OTA update, "
+                f"wait for otaproxy@{_upper_proxy} online..."
+            )
+            ensure_otaproxy_start(
+                _upper_proxy,
+                probing_timeout=WAIT_FOR_OTAPROXY_ONLINE,
+            )
+
```

Comment:
Good point! If this function timed out, it will raise a ConnectionError, and finally being caught by the outer caller(`_OTAUpdater.execute`), and the failure_reason(and optionally failure_traceback) will be recorded according to the raised exception.